File: 0101-0200/0137.Single_Number_II.js
//4ms



/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function(nums) {
    let map = new Map();
    let ans;
    for(let i=0;i<nums.length;i++){
        map.set(nums[i],(map.get(nums[i])||0)+1);
    }
    map.forEach((val,key)=>{
        if(val==1){
            ans = key;
        }
    })
    return ans;
};



//4ms same




/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function(nums) {
    let hashMap = new Map();
    for (let num of nums) {
        hashMap.set(num, (hashMap.get(num) || 0) + 1);
    }
    for (const [key,value] of hashMap) {
        if (hashMap.get(key) === 1) {
            return key;
        }
    }
};





//2ms







/**
 * Находит уникальный элемент в массиве, где все остальные встречаются трижды
 * @param {number[]} nums Входной массив чисел
 * @return {number} Уникальный элемент
 */
var singleNumber = function(nums) {
    let ones = 0, twos = 0;
    for (const num of nums) {
        // Добавляем в ones элементы, встретившиеся первый раз
        // и удаляем те, что уже есть в twos (т.е. встретились дважды)
        ones = (ones ^ num) & ~twos;
        
        // Добавляем в twos элементы, встретившиеся второй раз
        // и удаляем те, что уже есть в ones (т.е. встретились трижды)
        twos = (twos ^ num) & ~ones;
    }
    return ones;
};

File: 0101-0200/0137.Single_Number_II.rs
//0ms


impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
       
       let mut  num= nums;
num.sort();
        if num.len()==1{
            return num[0] as i32;
        }
        if num[num.len()-1]!=num[num.len()-2]{
            return num[num.len()-1] as i32;
        }
        for i in 0..num.len()-1{
            if i==0 && num[i]!=num[i+1]{
                return num[i] as i32;
            }
            if i>0 && num[i]!=num[i+1] && num[i]!=num[i-1]{
                return num[i] as i32;
            }
        }
        return 0;
    }
}
